File: problems/0167-two-sum-ii---input-array-is-sorted/solution.cpp
#include<unordered_map>

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        unordered_map<int, vector<int>> cache{}; // key: numVal, val: vector<idx>

        int i = 0;
        for (int num : numbers) {
            int diff = target - num;

            if (cache.find(diff) != cache.end()) {
                if (num == diff && cache[diff].size() >= 2) {
                    return {cache[diff][0] + 1, cache[diff][1] + 1}; // return 1st 2 idx from same key
                }
                return {cache[diff][0] + 1, i + 1}; // return idx and diff 1st idx
            }

            if (cache[num].size() == 0) {
                cache[num] = {i};
            } else {
                cache[num].push_back(i);
            }

            i++;
        }
        return {};
    }
};

File: problems/0167-two-sum-ii---input-array-is-sorted/solution.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        l = 0
        r = len(nums) - 1

        while l < r:
            total = nums[l] + nums[r]
            if total < target:
                l += 1
            elif total > target:
                r -= 1
            else:
                break
        
        return [l + 1, r + 1]
